src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServer.cs

```diff
@@ -183,6 +185,8 @@ private static ServiceContext CreateServiceContext(IOptions<KestrelServerOptions
             await Task.WhenAll(tasks).ConfigureAwait(false);
 
             _heartbeat.Dispose();
+
+            _stoppedTcs.TrySetResult(null);
```

Comment:
Instead of putting it in a finally, I'm catching any exception and completing the task with it and rethrowing.